File: clients/js/cascade.mjs
// Cascade client (JavaScript/TypeScript, zero deps — uses global fetch, Node >= 18 / Deno / Bun).
// Talks to a local `cascade gateway`.
//
//   import { Cascade } from "./cascade.mjs";
//   const c = new Cascade("http://127.0.0.1:7070");
//   await c.put("doc-1", "Turso has CDC + native replication", { src: "demo" });  // master gateway
//   for (const hit of await c.search("what does turso do?", 5))
//     console.log(hit.score.toFixed(3), hit.text);
//
// Run directly for a quick smoke test:  node cascade.mjs

export class Cascade {
  constructor(base = "http://127.0.0.1:7070") {
    this.base = base.replace(/\/$/, "");
  }
  async health() {
    return (await fetch(this.base + "/health")).json();
  }
  async put(id, text, meta = {}) {
    const r = await fetch(this.base + "/put", {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify({ id, text, meta }),
    });
    return r.json();
  }
  async search(q, k = 5) {
    const u = new URL(this.base + "/search");
    u.searchParams.set("q", q);
    u.searchParams.set("k", String(k));
    const r = await fetch(u);
    return (await r.json()).hits;
  }
  async drain() {
    return (await fetch(this.base + "/drain", { method: "POST" })).json();
  }
}

// Smoke test when run directly.
if (import.meta.url === `file://${process.argv[1]}`) {
  const c = new Cascade();
  console.log("health:", await c.health());
  await c.put("js-1", "Cascade exposes Turso over a local HTTP gateway", { src: "js" });
  for (const h of await c.search("how do clients talk to a node?", 3)) {
    console.log(`  ${h.score.toFixed(3)}  ${h.text}`);
  }
}

File: clients/go/cascade.go
// Package cascade is a tiny Go client for a local `cascade gateway` (stdlib only).
//
//	c := cascade.New("http://127.0.0.1:7070")
//	_ = c.Put("doc-1", "Turso has CDC + native replication", map[string]any{"src": "demo"})
//	hits, _ := c.Search("what does turso do?", 5)
//	for _, h := range hits { fmt.Printf("%.3f %s\n", h.Score, h.Text) }
package cascade

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"time"
)

type Client struct {
	Base string
	HTTP *http.Client
}

func New(base string) *Client {
	return &Client{Base: strings.TrimRight(base, "/"), HTTP: &http.Client{Timeout: 120 * time.Second}}
}

type Hit struct {
	ID    string          `json:"id"`
	Text  string          `json:"text"`
	Meta  json.RawMessage `json:"meta"`
	Score float64         `json:"score"`
}

func (c *Client) Health() (map[string]any, error) {
	var out map[string]any
	return out, c.do("GET", "/health", nil, &out)
}

func (c *Client) Put(id, text string, meta map[string]any) error {
	if meta == nil {
		meta = map[string]any{}
	}
	return c.do("POST", "/put", map[string]any{"id": id, "text": text, "meta": meta}, nil)
}

func (c *Client) Search(q string, k int) ([]Hit, error) {
	var out struct {
		Hits []Hit `json:"hits"`
	}
	path := "/search?" + url.Values{"q": {q}, "k": {strconv.Itoa(k)}}.Encode()
	return out.Hits, c.do("GET", path, nil, &out)
}

func (c *Client) Drain() (map[string]any, error) {
	var out map[string]any
	return out, c.do("POST", "/drain", nil, &out)
}

func (c *Client) do(method, path string, body any, out any) error {
	var rdr io.Reader
	if body != nil {
		b, _ := json.Marshal(body)
		rdr = bytes.NewReader(b)
	}
	req, err := http.NewRequest(method, c.Base+path, rdr)
	if err != nil {
		return err
	}
	if body != nil {
		req.Header.Set("content-type", "application/json")
	}
	resp, err := c.HTTP.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	data, _ := io.ReadAll(resp.Body)
	if resp.StatusCode >= 300 {
		return fmt.Errorf("cascade %s %s: %d %s", method, path, resp.StatusCode, string(data))
	}
	if out != nil {
		return json.Unmarshal(data, out)
	}
	return nil
}
